src/app/data-model/List.h

```diff
@@ -0,0 +1,87 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "Decode.h"
+#include "Encode.h"
+#include <iterator>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CodeUtils.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/*
+ * Dedicated type for list<T> that is at its base, just a Span.
+ *
+ * Motivated by the need to create distinction between Lists that use Spans
+ * vs. other data model types that use Spans (like octetstr). These have different
+ * encodings. Consequently, there needs to be an actual C++ type distinction to ensure
+ * correct specialization of the Encode/Decode methods.
+ *
+ */
+template <typename T>
+struct List : public Span<T>
+{
+    using Span<T>::Span;
```

Comment:
This is needed to make visible all constructors of the base class to appear within the derived class, as well as introduce functions in the base class into the derived class.

From cppreference.com:

> Using-declarations can be used to introduce namespace members into other namespaces and block scopes, or to introduce base class members into derived class definitions, or to introduce enumerators into namespaces, block, and class scopes (since C++20).

And

> If the using-declaration refers to a constructor of a direct base of the class being defined (e.g. using Base::Base;), all constructors of that base (ignoring member access) are made visible to overload resolution when initializing the derived class.
If overload resolution selects an inherited constructor, it is accessible if it would be accessible when used to construct an object of the corresponding base class: the accessibility of the using-declaration that introduced it is ignored.

Reply:
Will add a comment.

Reply:
Right, the operator= was clear; I did not realize you could `using` a base-class constructor and have it just work.  Today I learned.  ;)